File: job-scraper/scraper/sources/remoteok.py
"""RemoteOK — free public JSON API, no key required.
Docs: https://remoteok.com/api
"""

import requests

from ..models import Job, parse_epoch

API_URL = "https://remoteok.com/api"


def fetch(session: requests.Session, config: dict) -> list:
    headers = {"User-Agent": "daily-job-digest/1.0 (personal use; see README)"}
    resp = session.get(API_URL, headers=headers, timeout=20)
    resp.raise_for_status()
    data = resp.json()

    jobs = []
    for row in data:
        # First element of the response is a metadata/legal blob, not a job.
        if not isinstance(row, dict) or "id" not in row:
            continue
        jobs.append(
            Job(
                source="RemoteOK",
                title=row.get("position", "").strip(),
                company=row.get("company", "").strip(),
                url=row.get("url") or row.get("apply_url", ""),
                location=row.get("location", "").strip(),
                remote=True,
                posted_at=parse_epoch(row.get("epoch")),
                salary=_salary(row),
                tags=row.get("tags", []) or [],
            )
        )
    return jobs


def _salary(row: dict) -> str:
    lo, hi = row.get("salary_min"), row.get("salary_max")
    if lo and hi:
        return f"${lo:,} - ${hi:,}"
    return ""

File: job-scraper/src/sources/remoteok.mjs
// RemoteOK: free, unauthenticated public JSON feed (https://remoteok.com/api).
// Their license terms require attribution + a direct (non-redirected) link back
// to the listing, both of which the viewer/data output preserve.
export const name = "remoteok";

export function isConfigured() {
  return true; // no credentials required
}

export async function fetchJobs({ queries }) {
  const res = await fetch("https://remoteok.com/api", {
    headers: { "User-Agent": "job-scraper (+https://remoteok.com/api)" },
  });
  if (!res.ok) {
    console.warn(`[remoteok] fetch failed: ${res.status} ${res.statusText}`);
    return [];
  }
  const body = await res.json();
  // First element is a legal/metadata notice, not a job.
  const rows = Array.isArray(body) ? body.filter((r) => r && r.id) : [];

  // RemoteOK tags/titles are short ("frontend", "Interface Designer"), so match on
  // individual significant words from each query rather than the full phrase.
  const wordSets = queries.map((q) =>
    q
      .toLowerCase()
      .split(/\s+/)
      .filter((w) => w.length > 3)
  );
  return rows
    .filter((r) => {
      if (wordSets.length === 0) return true;
      const haystack = `${r.position ?? ""} ${(r.tags ?? []).join(" ")}`.toLowerCase();
      return wordSets.some((words) => words.some((w) => haystack.includes(w)));
    })
    .map((r) => ({
      id: `remoteok:${r.id}`,
      source: "RemoteOK",
      title: r.position,
      company: r.company,
      location: r.location || "Remote",
      url: r.url,
      postedAt: r.date,
      remote: true,
      salary: r.salary_min && r.salary_max ? `${r.salary_min}-${r.salary_max}` : null,
      description: (r.description ?? "").replace(/<[^>]+>/g, " ").slice(0, 280),
      tags: r.tags ?? [],
    }));
}
